### ISSUE TITLE
Lua on non-windows mono build

### ISSUE THREAD

**Issue by lsthiros** [TASVideos/BizHawk#1093]
On the mono-portable branch, Lua is currently marked as unsupported for non windows platforms. Upon trying a naive approach of removing the compatibility guard, the application will still crash after seemingly failing to create a working Lua instance. However, it is unclear to me whether this portability troubles are arising from the winform implementation, the Lua implementation, or something else entirely.

As per #921 and #1017 it seems as though the issue resides in the compatibility of the Lua engine on mono32.

Is this a correct assumption? What is preventing the use of Lua on the mono-portable build?

**Comment by Sappharad** [comment]
The last time I tried it, it was because the Lua library called native windows functions internally, despite being a managed DLL.

**Comment by theandrewk27** [comment]
Lua would not work on my system. I was told to open a Terminal and all the files would open in TextEdit.

**Comment by lsthiros** [comment]
@Sappharad
Which assembly holds the Lua implementation? I didn’t see any such target in the mono solution or elsewhere on the branch.

**Comment by Sappharad** [comment]
Check the separate LuaInterface directory for a project that does something. That builds LuaInterface.

It is referenced via the DLLs in References/x86/. Lua51.dll is a native windows DLL, not possible to support on Mono. You'd need the source code to that DLL and a build to replace it with a native macOS implementation. I'm not aware of source code existing that can be ported.

**Comment by lsthiros** [comment]
I’ll check out the DLL interface when I have the opportunity. I’m hoping that uses the usual 5.1 library functions as an entry point, but it’s possible that a wrapper was used and never documented.

**Comment by Sappharad** [comment]
Just keep in mind that whatever you do is only going to be useful on Linux going forward.

I stopped updating the mono-portable branch when 2.0 was released because the project is now 64-bit and WinForms is not supported for 64-bit targets on macOS and will never be. Since I was handling macOS, for that situation I'm slowly building a new UI from scratch in a fork on my GitHub account but that will only have the emulation part in common. There's no point in doing a 32-bit build for macOS anyway, since Apple notified developers last year that they plan to drop support for 32-bit apps for the OS that will release in Fall of 2019.

**Comment by lsthiros** [comment]
The UI and non-UI portions of Lua seem to be pretty decoupled.

Were you planning on keeping the same Lua functionalities, or a subset thereof,  in-tact with your UI rewrite?

**Comment by Sappharad** [comment]
I don't care about Lua, which is why it wasn't a priority on the previous macOS port. Once that's in a releasable state, someone else who wants it can add it.

**Comment by lsthiros** [comment]
Interesting to hear. I investigated LuaInterface and found the root compatibility issue. The managed/unmanaged interop occurs occurs between the LuaInterface assembly and Managed C++ in [LuaDLL.cpp](https://github.com/TASVideos/BizHawk/blob/mono-portable/LuaInterface/Lua/src/LuaDLL.cpp). Whatever the assembly ended up as, the interop is incompatible with mono. If a P/Invoke interface were used, it might be compatible.

I've made such interfaces before. I'll see what i can put together.

**Comment by vadosnaprimer** [comment]
There's also an option to cherry-pick commits that introduce NLua, over to the 1.x branch. I recently started cherry-picking LuaInterface fixes from there, so 1.x started moving somewhere, since I'm using it for my tas. NLua uses full C# reimplementation of lua, so it should work on mac (and linux).

**Comment by lsthiros** [comment]
Using NLua sounds like a great long-term compatibility solution on the 2.x developers part. However I do not know the difference between the 2.x and 1.x lua integrations which might be coupled differently with the other systems.

In the interest of keeping the issue limited to a mono compatibility fix, I’m going to stick with converting the Managed C++ interface to a P/Invoke interface for now. I’ll submit a pull request if I get it working (assuming that there is interest in this branch at all)

**Comment by Sappharad** [comment]
I have no objections to continuing the portable branch, if that's where you want the pull request for. There was also the guy who made a bunch of changes to fix linux, and his changes haven't been pulled back into here yet. He submitted a pull request at one point, but I mentioned some things about his initial changes and he basically re-did everything and it's still a fork right now. There are still mistakes though that he probably should fix if we ever pull that in though. For example, he's telling people to build MonoMacWrapper.sln for Linux, but I'm guessing he had no idea that MonoMac was the name of the legacy API to to implement native Mac UI's in Mono and that solution's purpose was to wrap the WinForms UI with native mac menus and dialogs. I have no idea how/why he's using any of that for Linux. I haven't gone through it recently to provide any feedback though.

**Comment by vadosnaprimer** [comment]
@lsthiros I'll merge the nlua commits within a week so you could just test it in the field.

**Comment by lsthiros** [comment]
I've hit a an issue in facilitating calls from lua functions to arbitrary C# functions. The P/Invoke port will add more complexity than I anticipated, but it won't be impossible. If @vadosnaprimer finished their workaround first, I'll just drop my work and test that out.

**Comment by lsthiros** [comment]
Getting closer to a full implementation. Everything compiles right now, but I need to add a large amount of testing before I'm sure that it will work.

I'm keeping my progress here:
https://github.com/lsthiros/BizHawk/tree/lua-mono-compat

**Comment by Chaneybenjamini** [comment]
Isthiros, Do you have build instructions for your fork?

**Comment by lsthiros** [comment]
@Chaneybenjamini Not yet, but there is a bare-bones makefile to build the Lua dylib, and a .sln to hold the C# code that worked on my machine. If you build those (and implement the luaL_loadbuffer function that I've putting off for some reason), you should have a LuaIntegration.dll assembly that will be syntactically compatible with the rest of the project.

However, I'm still writing unit tests, so I'm not going to offer any guarantees that the current code will work until they are finished.

**Comment by lsthiros** [comment]
<img width="1280" alt="mario-lua" src="https://user-images.githubusercontent.com/3664023/35180098-16cda3a2-fd5d-11e7-8a12-f3d420280379.png">
Finished basic unit tests, but I'd like to get some code coverage metrics before I continue them.
Also, need to integrate it into the build system, but the commit I have on my fork should work if you correct the references to LuaInterface and add a reference to LuaInterfaceWrapper in the main sln.

**Comment by vadosnaprimer** [comment]
Awesome job! I didn't have free time until weekend. Will try today.

**Comment by theandrewk27** [comment]
Looks awesome. Is this on OSX 10.13.2?

**Comment by lsthiros** [comment]
@theandrewk27 Yes

**Comment by theandrewk27** [comment]
Great. This will make things a lot easier because I couldn't run anything.

**Comment by polytomous** [comment]
Do we have any existing integration tests for lua use cases? I've made a few trivial (read: static analysis fixes and resharper stuff) changes there in a fork and am curious what testing we have in place there. I didn't notice any unit testing, but maybe that's in a different project.

**Comment by vadosnaprimer** [comment]
Ask @zeromus.

**Comment by lsthiros** [comment]
The feature that I was targeting, Lua support on mono, is functionally complete. What would be the integration requirements before merging it with the mono-portable branch?

**Comment by zeromus** [comment]
that it not break building on linux, if it previously built in linux; and not break building on mac, if it previously built on mac?

**Comment by Chaneybenjamini** [comment]
xbuild BizHawk.sln is still failing for me. Should I be doing something different?

**Comment by lsthiros** [comment]
Did you target the x86 Debug build? What was the error? Also, if this is a build issue concerning my branch, please post it on my fork.

**Comment by Chaneybenjamini** [comment]
I believe that it is an issue with your branch (or more likely my use of your branch). For some reason I can't figure out how to post on Issue on your fork. Is it possible they are disabled?

**Comment by lsthiros** [comment]
They were disabled. Didn’t know that was the default setting 😕
Either way they’re enabled now.

**Comment by vadosnaprimer** [comment]
@lsthiros can you make a PR? I failed to do my part of this, but now that 1.x branch has its own releases, its mono port will benefit a lot from lua I guess.

**Comment by lsthiros** [comment]
Wow, I pretty much lost track of this project over the past couple months. I remember being blocked on [code coverage metrics for DotNet Core](https://github.com/Microsoft/vstest/issues/981), which has since been added.
The build on my branch should still be functional, so I'll look into the complexities of a 1.x merge. Granted, the last time I tried using my changes, the performance was horrendous for reasons that I could not isolate to any one branch.
